extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -220,12 +203,6 @@ func (a *genericActuator) Reconcile(ctx context.Context, log logr.Logger, worker
 		return fmt.Errorf("failed to update the machine deployments in worker status: %w", err)
 	}
 
-	// Cleanup machine dependencies.
-	// TODO(dkistner): Remove in a future release.
-	if err := workerDelegate.CleanupMachineDependencies(ctx); err != nil {
-		return fmt.Errorf("failed to cleanup machine dependencies: %w", err)
-	}
```

Comment:
For the funcs added with https://github.com/gardener/gardener/pull/6290, I am not sure if provider-azure adapted to the new funcs. Ref https://github.com/gardener/gardener-extension-provider-azure/blob/d31fce706f474b8b1fe211dc22a882f3ba23a762/pkg/controller/worker/machine_dependencies.go#L45-L65. Hence, I think first provider-azure and any other extension relying on the old deprecated funcs has to be adapted.
Only then we should be able to drop them. @dkistner correct me if I am wrong.

Reply:
/hold
until the above comment is clarified

Reply:
`provider-azure` (and others still implementing the removed functions) can simply adapt when revendoring the next `gardener/gardener` version since the change is in the extensions library. There is no need to do this upfront.
/unhold

Reply:
I had a look into the required provider-azure adaptation. I drafted locally https://github.com/ialidzhikov/gardener-extension-provider-azure/commit/0101e56ef806de4c9129f5e105353995a8b46f50. The adaptation for `DeployMachineDependencies` is easy - the logic in `DeployMachineDependencies` has to be moved to `PreReconcileHook`. The adaptation for `CleanupMachineDependencies` is to use the  `PostReconcileHook` and `PostDeleteHook` funcs. Adaptation that would work for now would be to replicate the behaviour of the existing `CleanupMachineDependencies` to the new `PostReconcileHook` and `PostDeleteHook` funcs. Having a look into https://github.com/gardener/gardener/pull/6290 and the provider-azure code, I think the motivation to replace `CleanupMachineDependencies` in favor of `PostReconcileHook` and `PostDeleteHook` funcs is to do different things post-reconcile and post-delete (so far the same `CleanupMachineDependencies` was called after both). In short, the adaptation in https://github.com/ialidzhikov/gardener-extension-provider-azure/commit/0101e56ef806de4c9129f5e105353995a8b46f50 would work but having a look into provider-azure code it can be further improved/optimised.

I see that provider-openstack has to be adapted as well - ref https://github.com/gardener/gardener-extension-provider-openstack/blob/1927d05135160571ab746ac5ee263ba45119e4c7/pkg/controller/worker/machine_dependencies.go.

> `provider-azure` (and others still implementing the removed functions) can simply adapt when revendoring the next `gardener/gardener` version since the change is in the extensions library. There is no need to do this upfront.

Yes, technically such adaptation can be done while revondoring. But it would be yet another thing to take care of when revendoring these extensions. It can be tackled separately now and make the revendoring of new g/g version easier. IMO removing the deprecated things should be the very last step when we are sure that usages from external components are adapted properly.
I would rather vote to first adapt the usages in the extensions (maybe even help the colleagues with this).

---

@dkistner @kon-angelo for provider-azure do you want me to create a PR with the minimal adaptation described above or do you want to take care with more extensive adaptation/refactoring?
@dkistner @kon-angelo @MartinWeindel for provider-openstack, do you want me to check or will you take care?

Reply:
I think this was discussed multiple times already, but anyways, let's do it again: The proposed approach does neither scale nor work. We don't know which extensions exist out in the wild which might still implement these functions (like `provider-azure` or `provider-openstack`), and we won't go hunting for them (some of them might not even be open-source).

The whole purpose of first deprecating functionality without removing it at once is to give dependants time to adapt to the incoming changes. We can't help if this is ignored/missed/not done for whatever reasons. At some point in time we have to drop the code.
Sometimes we take over this work for extensions we know of (for good will or efficiency reasons, I'd say), but in general this is not our responsibility, it's that of the maintainers of dependent components.
Taking over this work is also neither better nor worse in the context of the deprecation process. Instead, we should have clear policies (e.g., when functionality is deprecated with version `x` then it is removed after three releases with version `x+3`). I think since recently we try to follow this approach more consistently.

To sum it up: I don't see a reason to artificially delay this even further now (it's deprecated for almost a year already). We can see the effects if we don't clean up deprecated functionality (code is not adapted, hence technical debt is carried along for too long). While it is not particularly complex/problematic in this specific case (the code doesn't bother us), I still don't think that this instance should be treated in a special way.

Reply:
I now created minimal adaptation PRs for provider-azure and provider-openstack - https://github.com/gardener/gardener-extension-provider-azure/pull/645 and https://github.com/gardener/gardener-extension-provider-openstack/pull/584.